### ISSUE TITLE
Containerized veos instructions broken using centos:latest

### ISSUE THREAD

**Issue by ronmcmillian** [networkop/docker-topo#35]
When following the instructions for creating a containerized veos, the build fails. This is due to changes in CentOS 8 such as this one: https://forums.centos.org/viewtopic.php?t=71970. I was able to get the image to build by changing centos:latest in the Dockerfile to centos:7.

**Comment by networkop** [comment]
yep, looks like a problem. Can you do a PR?

**Comment by ronmcmillian** [comment]
I can after I get it working. After following the directions, should I be able to plug  the new image name in my .yml and rebuild? 

It's currently failing with "docker.errors.APIError: 400 Client Error: Bad Request ("OCI runtime create failed: container_linux.go:346: starting container process caused "exec: \"/sbin/init\": stat /sbin/init: no such file or directory": unknown")".

**Comment by ronmcmillian** [comment]
actually, I see it looks to be based on device name.

**Comment by networkop** [comment]
yeah, device name is what decides which image to use. so veos needs to be prefixed with "veos".
alternatively, you can define your own prefix->image mapping with `custom_image`

**Comment by ronmcmillian** [comment]
I have renamed my devices to veos_<device name>. I'm now hitting a new error. Any ideas? 

```DEBUG:__main__:	Starting container rc2_veos_ch2-csw-01
DEBUG:__main__:	Obtaining a pointer to container rc2_veos_ch2-csw-01
DEBUG:urllib3.connectionpool:http://localhost:None "GET /v1.35/containers/rc2_veos_ch2-csw-01/json HTTP/1.1" 404 53
DEBUG:urllib3.connectionpool:http://localhost:None "POST /v1.35/containers/create?name=rc2_veos_ch2-csw-01 HTTP/1.1" 400 35
Traceback (most recent call last):
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/api/client.py", line 261, in _raise_for_status
    response.raise_for_status()
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http+docker://localhost/v1.35/containers/create?name=rc2_veos_ch2-csw-01

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/home/core/network-lab/docker-topo/testdir/bin/docker-topo", line 840, in <module>
    main()
  File "/var/home/core/network-lab/docker-topo/testdir/bin/docker-topo", line 775, in main
    started = [device.start() == 0 for (name, device) in devices.items()]
  File "/var/home/core/network-lab/docker-topo/testdir/bin/docker-topo", line 775, in <listcomp>
    started = [device.start() == 0 for (name, device) in devices.items()]
  File "/var/home/core/network-lab/docker-topo/testdir/bin/docker-topo", line 339, in start
    self._get_or_create()
  File "/var/home/core/network-lab/docker-topo/testdir/bin/docker-topo", line 306, in _get_or_create
    self.container = self._create()
  File "/var/home/core/network-lab/docker-topo/testdir/bin/docker-topo", line 326, in _create
    tty=True
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/models/containers.py", line 861, in create
    resp = self.client.api.create_container(**create_kwargs)
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/api/container.py", line 430, in create_container
    return self.create_container_from_config(config, name)
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/api/container.py", line 441, in create_container_from_config
    return self._result(res, True)
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/api/client.py", line 267, in _result
    self._raise_for_status(response)
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/api/client.py", line 263, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/var/home/core/network-lab/docker-topo/testdir/lib/python3.6/site-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.APIError: 400 Client Error: Bad Request ("No command specified")

**Comment by networkop** [comment]
hm.. strange. it seems to complain that the `command` argument is missing (as in `self.command`). can you share you topology yaml?

**Comment by ronmcmillian** [comment]
Sure. 

```VERSION: 2
driver: veth
CONF_DIR: './config'
CEOS_IMAGE: veos:latest

links:
# link 1, multipoint for mgmt network
  - endpoints: ["veos_ch2-csw-01:Management1", "veos_ch2-csw-02:Management1", "veos_ch2-asw-01:Management1", "veos_ch2-asw-02:Management1", "veos_ch2-psw-01:Management1", "veos_ch2-psw-02:Management1", "veos_ny4-csw-01:Management1", "veos_ny4-csw-02:Management1", "veos_ny4-asw-01:Management1", "veos_ny4-asw-02:Management1", "veos_ny4-psw-01:Management1", "veos_ny4-psw-02:Management1"]
    driver: bridge
# veos_ch2-csw-01 
  - endpoints: ["veos_ch2-csw-01:eth1", "veos_ny4-csw-01:eth1"]
  - endpoints: ["veos_ch2-csw-01:eth49", "veos_ch2-asw-01:eth49"]
  - endpoints: ["veos_ch2-csw-01:eth50", "veos_ch2-asw-02:eth49"]
  - endpoints: ["veos_ch2-csw-01:eth51", "veos_ch2-psw-01:eth13"]
  - endpoints: ["veos_ch2-csw-01:eth52", "veos_ch2-psw-02:eth13"]

# veos_ch2-csw-02
  - endpoints: ["veos_ch2-csw-02:eth49", "veos_ch2-asw-01:eth50"]
  - endpoints: ["veos_ch2-csw-02:eth50", "veos_ch2-asw-02:eth50"]
  - endpoints: ["veos_ch2-csw-02:eth51", "veos_ch2-psw-01:eth14"]
  - endpoints: ["veos_ch2-csw-02:eth52", "veos_ch2-psw-02:eth14"]

# asw-0x
  - endpoints: ["veos_ch2-asw-01:eth53", "veos_ch2-asw-02:eth53"]
  - endpoints: ["veos_ch2-asw-01:eth54", "veos_ch2-asw-02:eth54"]

# veos_ny4-csw-01
  - endpoints: ["veos_ny4-csw-01:eth49", "veos_ny4-asw-01:eth49"]
  - endpoints: ["veos_ny4-csw-01:eth50", "veos_ny4-asw-02:eth49"]
  - endpoints: ["veos_ny4-csw-01:eth51", "veos_ny4-psw-01:eth13"]
  - endpoints: ["veos_ny4-csw-01:eth52", "veos_ny4-psw-02:eth13"]

# veos_ny4-csw-02
  - endpoints: ["veos_ny4-csw-02:eth49", "veos_ny4-asw-01:eth50"]
  - endpoints: ["veos_ny4-csw-02:eth50", "veos_ny4-asw-02:eth50"]
  - endpoints: ["veos_ny4-csw-02:eth51", "veos_ny4-psw-01:eth14"]
  - endpoints: ["veos_ny4-csw-02:eth52", "veos_ny4-psw-02:eth14"]

  # asw-0x
  - endpoints: ["veos_ny4-asw-01:eth54", "veos_ny4-asw-02:eth53"]
  - endpoints: ["veos_ny4-asw-01:eth53", "veos_ny4-asw-02:eth54"]
```

**Comment by networkop** [comment]
i think you're missing `VERSION: 2` in your yaml. can you try adding that to the top and running again?

**Comment by networkop** [comment]
similar to https://github.com/networkop/docker-topo/blob/master/topo-extra-files/examples/v2/5-node-ls.yml#L11

**Comment by ronmcmillian** [comment]
Sorry, I messed up the formatting when pasting earlier. 'VERSION: 2' is already in my topology file. 

```
VERSION: 2
driver: veth
CONF_DIR: './config'
CEOS_IMAGE: veos:latest

links:
# link 1, multipoint for mgmt network
  - endpoints: ["veos_ch2-csw-01:Management1", "veos_ch2-csw-02:Management1", "veos_ch2-asw-01:Management1", "veos_ch2-asw-02:Management1", "veos_ch2-psw-01:Management1", "veos_ch2-psw-02:Management1", "veos_ny4-csw-01:Management1", "veos_ny4-csw-02:Management1", "veos_ny4-asw-01:Management1", "veos_ny4-asw-02:Management1", "veos_ny4-psw-01:Management1", "veos_ny4-psw-02:Management1"]
    driver: bridge
# veos_ch2-csw-01 
  - endpoints: ["veos_ch2-csw-01:eth1", "veos_ny4-csw-01:eth1"]
  - endpoints: ["veos_ch2-csw-01:eth49", "veos_ch2-asw-01:eth49"]
  - endpoints: ["veos_ch2-csw-01:eth50", "veos_ch2-asw-02:eth49"]
  - endpoints: ["veos_ch2-csw-01:eth51", "veos_ch2-psw-01:eth13"]
  - endpoints: ["veos_ch2-csw-01:eth52", "veos_ch2-psw-02:eth13"]

# veos_ch2-csw-02
  - endpoints: ["veos_ch2-csw-02:eth49", "veos_ch2-asw-01:eth50"]
  - endpoints: ["veos_ch2-csw-02:eth50", "veos_ch2-asw-02:eth50"]
  - endpoints: ["veos_ch2-csw-02:eth51", "veos_ch2-psw-01:eth14"]
  - endpoints: ["veos_ch2-csw-02:eth52", "veos_ch2-psw-02:eth14"]

# asw-0x
  - endpoints: ["veos_ch2-asw-01:eth53", "veos_ch2-asw-02:eth53"]
  - endpoints: ["veos_ch2-asw-01:eth54", "veos_ch2-asw-02:eth54"]

# veos_ny4-csw-01
  - endpoints: ["veos_ny4-csw-01:eth49", "veos_ny4-asw-01:eth49"]
  - endpoints: ["veos_ny4-csw-01:eth50", "veos_ny4-asw-02:eth49"]
  - endpoints: ["veos_ny4-csw-01:eth51", "veos_ny4-psw-01:eth13"]
  - endpoints: ["veos_ny4-csw-01:eth52", "veos_ny4-psw-02:eth13"]

# veos_ny4-csw-02
  - endpoints: ["veos_ny4-csw-02:eth49", "veos_ny4-asw-01:eth50"]
  - endpoints: ["veos_ny4-csw-02:eth50", "veos_ny4-asw-02:eth50"]
  - endpoints: ["veos_ny4-csw-02:eth51", "veos_ny4-psw-01:eth14"]
  - endpoints: ["veos_ny4-csw-02:eth52", "veos_ny4-psw-02:eth14"]

  # asw-0x
  - endpoints: ["veos_ny4-asw-01:eth54", "veos_ny4-asw-02:eth53"]
  - endpoints: ["veos_ny4-asw-01:eth53", "veos_ny4-asw-02:eth54"]

**Comment by networkop** [comment]
I can't run your full topo on my laptop but the top 4-5 devices work fine. I think there's something messed up with the veos image. A couple of questions

1. Are you able to run veos:latest image as a standalone container. e.g.
```
echo "hostname MYRANDOM-HOSTNAME" > my-config
docker run -d --name veos -v $(pwd)/my-config:/mnt/flash/startup-config --privileged veos
docker exec -it veos bash
[root@de16326e92f2 /]# telnet localhost 23
MYRANDOM-HOSTNAME login:      
```

2. how does your `docker image inspect veos:latest` look like?

**Comment by ronmcmillian** [comment]
hrm. No, I get an error when running the docker run command. 

`(testdir) core@localhost:~/network-lab/docker-topo/temp$ docker run -d --name veos -v $(pwd)/my-config:/mnt/flash/startup-config --privileged veos
docker: Error response from daemon: No command specified.
See 'docker run --help'.`

2. Please see below: 

```(testdir) core@localhost:~/network-lab/docker-topo/veos$ docker image inspect veos:latest
[
    {
        "Id": "sha256:06adc86196a0236d7028bf0aae6151d92b8851b9021dcd070b798f3b4586c376",
        "RepoTags": [
            "veos:4.22.2M",
            "veos:latest"
        ],
        "RepoDigests": [],
        "Parent": "",
        "Comment": "Imported from -",
        "Created": "2020-02-19T18:11:54.710128969Z",
        "Container": "",
        "ContainerConfig": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "DockerVersion": "18.09.8",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 1684990687,
        "VirtualSize": 1684990687,
        "GraphDriver": {
            "Data": {
                "MergedDir": "/var/lib/docker/overlay2/af0d8f3cb3006b86ddb300b94536ab0f2a9aec1356550d8312f8611e0eef09de/merged",
                "UpperDir": "/var/lib/docker/overlay2/af0d8f3cb3006b86ddb300b94536ab0f2a9aec1356550d8312f8611e0eef09de/diff",
                "WorkDir": "/var/lib/docker/overlay2/af0d8f3cb3006b86ddb300b94536ab0f2a9aec1356550d8312f8611e0eef09de/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:300b8dc54c533a2ec4e897443ca69e15da8823e86e9834ac142eed72004e6395"
            ]
        },
        "Metadata": {
            "LastTagTime": "2020-02-19T18:12:47.680758997Z"
        }
    }
]`

**Comment by ronmcmillian** [comment]
There seems to be something wrong with my image. I'm currently looking into it. Please standby

Edit: looks like when I moved the image, I lost a lot of details: https://serverfault.com/questions/757210/no-command-specified-from-re-imported-docker-image-container

**Comment by networkop** [comment]
yeah, it looks like the image was not built correctly. the entrypoint( which should be entrypoint.sh) is missing. did you run into any issues when doing docker build? (apart from the incorrect base image).

my suggestion would be to `docker image rm veos:latest` and rebuild it again. you should see the following 8 steps:
```
docker build -t veos .
Sending build context to Docker daemon  439.4MB
Step 1/8 : FROM centos:7
 ---> 9f38484d220f
Step 2/8 : RUN yum -y install epel-release &&     yum makecache fast &&     yum install -y qemu-kvm bridge-utils iproute libvirt libvirt-client genisoimage virt-install telnet &&     yum clean all
 ---> Using cache
 ---> 2c83bd6ee606
Step 3/8 : COPY veos.qcow2 /var/lib/libvirt/images/
 ---> Using cache
 ---> 58885b05518d
Step 4/8 : COPY entrypoint.sh /
 ---> Using cache
 ---> ddb272dc6da5
Step 5/8 : RUN mkdir -p /mnt/flash
 ---> Using cache
 ---> 0d2c85e04449
Step 6/8 : COPY rc.eos /mnt/flash/rc.eos
 ---> Using cache
 ---> 7d1ff3334872
Step 7/8 : RUN chmod +x /entrypoint.sh
 ---> Using cache
 ---> b2a2597e9cde
Step 8/8 : ENTRYPOINT /entrypoint.sh
 ---> Using cache
 ---> 40d4b1cfee3e
Successfully built 40d4b1cfee3e
Successfully tagged veos:latest
```

**Comment by networkop** [comment]
ah ok, makes sense. if you import /export it, you'd need to reset the entrypoint as it gets lost.
see https://github.com/networkop/cvp-in-docker for an example.

**Comment by ronmcmillian** [comment]
I rebuilt the image and tried again. It seems like the image is stopping within 10 seconds of starting  after running `docker run -d --name veos --privileged veos`

My `docker image inspect veos` is below. 

```
(testdir) core@localhost:~/docker-topo/topo-extra-files/veos$ docker image inspect veos
[
    {
        "Id": "sha256:61efb2c4a263a12b386b58226bd5792b24b457ece007e55b0c7b563f8e3e6294",
        "RepoTags": [
            "veos:latest"
        ],
        "RepoDigests": [],
        "Parent": "sha256:b68e81edb4de3bca3c94116f83e9aab8d95b5a6848b96445634b2c8ca6b8da4d",
        "Comment": "",
        "Created": "2020-02-19T21:21:48.912787741Z",
        "Container": "78ad9b13340a04faa763ce0264cfa493b74e0e1d1a32fae3d57268c9994e7d89",
        "ContainerConfig": {
            "Hostname": "78ad9b13340a",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(nop) ",
                "ENTRYPOINT [\"/bin/sh\" \"-c\" \"/entrypoint.sh\"]"
            ],
            "ArgsEscaped": true,
            "Image": "sha256:b68e81edb4de3bca3c94116f83e9aab8d95b5a6848b96445634b2c8ca6b8da4d",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/bin/sh",
                "-c",
                "/entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {
                "org.label-schema.build-date": "20191001",
                "org.label-schema.license": "GPLv2",
                "org.label-schema.name": "CentOS Base Image",
                "org.label-schema.schema-version": "1.0",
                "org.label-schema.vendor": "CentOS"
            }
        },
        "DockerVersion": "18.09.8",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": null,
            "ArgsEscaped": true,
            "Image": "sha256:b68e81edb4de3bca3c94116f83e9aab8d95b5a6848b96445634b2c8ca6b8da4d",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/bin/sh",
                "-c",
                "/entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {
                "org.label-schema.build-date": "20191001",
                "org.label-schema.license": "GPLv2",
                "org.label-schema.name": "CentOS Base Image",
                "org.label-schema.schema-version": "1.0",
                "org.label-schema.vendor": "CentOS"
            }
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 838476740,
        "VirtualSize": 838476740,
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/25ef8fdd50fdad9be4c58ecdd4edc076d6de0a67a3ba40e5be9efacaa661a184/diff:/var/lib/docker/overlay2/809675f013b46541a08f6fc3b419854fde42d6ab71556d03640823af91f85683/diff:/var/lib/docker/overlay2/289841181e7416c97f128145d7c7630539de3234ea862b0e6673d361429dfd4a/diff:/var/lib/docker/overlay2/36977077e08f589d8caecd554b17b1faecfac7d3eaa7e910d29c0fd8187bf23b/diff:/var/lib/docker/overlay2/ae585d784244b19a4c011aa8d28ebff771591e70ab618cccbb49dddfe9778cf1/diff:/var/lib/docker/overlay2/2870cdaab62ec30c9b5e31f893aaa06a61b8238d8d85ea29a2e4cef29201f878/diff",
                "MergedDir": "/var/lib/docker/overlay2/b4612809dd3a5bd6210a4e30bee75d28d7595716ff13ab3e7beec1cd2d24aa93/merged",
                "UpperDir": "/var/lib/docker/overlay2/b4612809dd3a5bd6210a4e30bee75d28d7595716ff13ab3e7beec1cd2d24aa93/diff",
                "WorkDir": "/var/lib/docker/overlay2/b4612809dd3a5bd6210a4e30bee75d28d7595716ff13ab3e7beec1cd2d24aa93/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:77b174a6a187b610e4699546bd973a8d1e77663796e3724318a2a4b24cb07ea0",
                "sha256:48a5124418fda2fea09d89f02c05f3716ef89546853a35b2ff0be8719a056c0c",
                "sha256:b47bd41c6aee099680d0dd492f5a8b515797b18af778c31149d6aa96a3f7579d",
                "sha256:3f1e665d3a2d17a52f00379bb33f3b4e15d86672762aff9010fd9d4b0db04ea9",
                "sha256:3d8addaf524675361a13de3fb4a1781599bc2b00b89321fc72d57d28c0cfd140",
                "sha256:10774d7cf0318caca3f4192ff53caa3073c25725e909d4a9ecd2e871fceddf50",
                "sha256:9ef839eb7e0fb3055859b4edc93cbfdc1f9bc3072c7a279a9894be52ab4adc0f"
            ]
        },
        "Metadata": {
            "LastTagTime": "2020-02-19T21:21:49.021705166Z"
        }
    }
]

**Comment by networkop** [comment]
what do you see in the "docker logs veos" ?

**Comment by ronmcmillian** [comment]
It looks like I'm having a kvm dependency issue. I think it's due to the VM that I'm running docker on. I don't think my issues now are related to your code. I'll get that PR in soon.

**Comment by ronmcmillian** [comment]
**Edit: This was a local issue for me related to kvm extensions and VT-x.** 

I receive the following from `docker logs veos` which I assumed to mean I was missing a kvm dependency. Looking at it again this doesn't seem to be the case. Have you seen this before? 

```Sleeping to wait for all interfaces to be connected
#####################
# Checking /dev/kvm #
#####################
Making sure that character device /dev/kvm exists and setting the right permissions
Requirement not satisfied: /dev/kvm not present

**Comment by ronmcmillian** [comment]
As background, this is to workaround https://github.com/networkop/docker-topo/issues/29

**Comment by networkop** [comment]
ok, this looks like you don't have nested (or hardware-assisted)
virtualisation enabled for your VM. It is what Intel calls VT-x, can't
remember what AMD calls it.
once you enable it, you should see the /dev/kvm device in your host os and
inside the container as well.

**Comment by networkop** [comment]
i believe #29 may be resolved now. i haven't checked myself but I remember seeing it somewhere that cEOS now supports dot1q trunking in recent releases. I might check if it's true in the next few days.

**Comment by networkop** [comment]
yeah, it looks like #29 has been solved since 4.23.1F onwards.

**Comment by ronmcmillian** [comment]
Ah, sounds good. I'll try an upgrade. I think this issue is set as well.
